### ISSUE TITLE
--watch not working due to permission denied error

### ISSUE THREAD

**Issue by nroi** [k0kubun/xremap#59]
First, thank you for this awesome tool!

I've tried the `--watch` option, and it doesn't seem to work as intended on my machine. I've [forked](https://github.com/nroi/xremap/tree/debugging-statements) the repo and added a few println! statements to understand what's going on, so here's what I did to reproduce the issue with the forked version:

1. Start `./target/release/xremap --watch /path/to/config.yml`. So far, everything's working as intended: It correctly detects `/dev/event3` as my keyboard. This is the output:
    ```
    Path "/dev/input/event4" cannot be opened: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
    Selecting devices from the following list:
    ------------------------------------------------------------------------------
    /dev/input/event0 : Power Button
    /dev/input/event1 : Power Button
    /dev/input/event2 : USB Keyboard Consumer Control
    /dev/input/event3 : USB Keyboard
    /dev/input/event5 : UVC Camera (046d:0825)
    /dev/input/event6 : PixArt USB Optical Mouse
    /dev/input/event7 : HDA ATI HDMI HDMI/DP,pcm=3
    /dev/input/event8 : HDA ATI HDMI HDMI/DP,pcm=7
    /dev/input/event9 : HDA ATI HDMI HDMI/DP,pcm=8
    /dev/input/event10: HDA ATI HDMI HDMI/DP,pcm=9
    /dev/input/event11: HDA ATI HDMI HDMI/DP,pcm=10
    /dev/input/event12: HDA ATI HDMI HDMI/DP,pcm=11
    /dev/input/event13: HD-Audio Generic Front Mic
    /dev/input/event14: HD-Audio Generic Rear Mic
    /dev/input/event15: HD-Audio Generic Line
    /dev/input/event16: HD-Audio Generic Line Out Front
    /dev/input/event17: HD-Audio Generic Line Out Surround
    /dev/input/event18: HD-Audio Generic Line Out CLFE
    /dev/input/event19: HD-Audio Generic Front Headphone
    /dev/input/event20: BurrBrown from Texas Instruments USB AUDIO  CODEC
    /dev/input/event21: USB Keyboard System Control
    ------------------------------------------------------------------------------
    Selected keyboards automatically since --device options weren't specified:
    ------------------------------------------------------------------------------
    /dev/input/event3 : USB Keyboard
    ------------------------------------------------------------------------------
    Received events from inotify: Err(EAGAIN)
    ```

2. Disconnect the keyboard. This is the output:

    ```
    Received events from inotify: Ok([InotifyEvent { wd: WatchDescriptor { wd: 1 },
    mask: IN_ATTRIB, cookie: 0, name: Some("event3") }])
    Detected device changes. Reselecting devices.
    Path "/dev/input/event2" cannot be opened: Os { code: 13, kind:
    PermissionDenied, message: "Permission denied" }
    Selecting devices from the following list:
    ------------------------------------------------------------------------------
    /dev/input/event0 : Power Button
    /dev/input/event1 : Power Button
    /dev/input/event5 : UVC Camera (046d:0825)
    /dev/input/event6 : PixArt USB Optical Mouse
    /dev/input/event7 : HDA ATI HDMI HDMI/DP,pcm=3
    /dev/input/event8 : HDA ATI HDMI HDMI/DP,pcm=7
    /dev/input/event9 : HDA ATI HDMI HDMI/DP,pcm=8
    /dev/input/event10: HDA ATI HDMI HDMI/DP,pcm=9
    /dev/input/event11: HDA ATI HDMI HDMI/DP,pcm=10
    /dev/input/event12: HDA ATI HDMI HDMI/DP,pcm=11
    /dev/input/event13: HD-Audio Generic Front Mic
    /dev/input/event14: HD-Audio Generic Rear Mic
    /dev/input/event15: HD-Audio Generic Line
    /dev/input/event16: HD-Audio Generic Line Out Front
    /dev/input/event17: HD-Audio Generic Line Out Surround
    /dev/input/event18: HD-Audio Generic Line Out CLFE
    /dev/input/event19: HD-Audio Generic Front Headphone
    /dev/input/event20: BurrBrown from Texas Instruments USB AUDIO  CODEC
    /dev/input/event21: USB Keyboard System Control
    ------------------------------------------------------------------------------
    Selected keyboards automatically since --device options weren't specified:
    ------------------------------------------------------------------------------
    warning: No device was selected, but --watch is waiting for new devices.
    ------------------------------------------------------------------------------
    ```

3. Reconnect the keyboard. Notice that in the output, `event2` and `event3` (the keyboard) cannot be opened (`PermissionDenied`).
    ```
    Received events from inotify: Ok([InotifyEvent { wd: WatchDescriptor { wd: 1 }, mask: IN_CREATE, cookie: 0, name: Some("event3") }, InotifyEvent { wd: WatchDescriptor { wd: 1 }, mask: IN_ATTRIB, cookie: 0, name: Some("event3") }])
    Detected device changes. Reselecting devices.
    Path "/dev/input/event21" cannot be opened: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
    Path "/dev/input/event4" cannot be opened: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
    Path "/dev/input/event2" cannot be opened: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
    Path "/dev/input/event3" cannot be opened: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
    Selecting devices from the following list:
    ------------------------------------------------------------------------------
    /dev/input/event0 : Power Button
    /dev/input/event1 : Power Button
    /dev/input/event5 : UVC Camera (046d:0825)
    /dev/input/event6 : PixArt USB Optical Mouse
    /dev/input/event7 : HDA ATI HDMI HDMI/DP,pcm=3
    /dev/input/event8 : HDA ATI HDMI HDMI/DP,pcm=7
    /dev/input/event9 : HDA ATI HDMI HDMI/DP,pcm=8
    /dev/input/event10: HDA ATI HDMI HDMI/DP,pcm=9
    /dev/input/event11: HDA ATI HDMI HDMI/DP,pcm=10
    /dev/input/event12: HDA ATI HDMI HDMI/DP,pcm=11
    /dev/input/event13: HD-Audio Generic Front Mic
    /dev/input/event14: HD-Audio Generic Rear Mic
    /dev/input/event15: HD-Audio Generic Line
    /dev/input/event16: HD-Audio Generic Line Out Front
    /dev/input/event17: HD-Audio Generic Line Out Surround
    /dev/input/event18: HD-Audio Generic Line Out CLFE
    /dev/input/event19: HD-Audio Generic Front Headphone
    /dev/input/event20: BurrBrown from Texas Instruments USB AUDIO  CODEC
    ------------------------------------------------------------------------------
    Selected keyboards automatically since --device options weren't specified:
    ------------------------------------------------------------------------------
    warning: No device was selected, but --watch is waiting for new devices.
    ------------------------------------------------------------------------------
    ```

So to my understanding, this is a race condition where the USB device is not yet readable immediately after the inotify event. According to the inotify println! statements that I've added, we receive an `IN_ATTRIB` event in addition to the `IN_CREATE` event when the device is reconnected. So apparently, even after the first `IN_ATTRIB` event, the device is still not readable.

I ran `notifywait -r -m /dev/input | grep 'event3'` to see what events occur after the keyboard is connected, here's the output:
```
/dev/input/ CREATE event3
/dev/input/ ATTRIB event3
/dev/input/ OPEN event3
/dev/input/ CLOSE_NOWRITE,CLOSE event3
/dev/input/ OPEN event3
/dev/input/ ATTRIB event3
/dev/input/ ATTRIB event3
/dev/input/ ATTRIB event3
/dev/input/ CLOSE_NOWRITE,CLOSE event3
/dev/input/ OPEN event3
/dev/input/ CLOSE_WRITE,CLOSE event3
/dev/input/ OPEN event3
/dev/input/ CLOSE_WRITE,CLOSE event3
/dev/input/ OPEN event3
/dev/input/ MODIFY event3
```

I really have no idea why I get so many `ATTRIB` events, and why it closes the file multiple times after the device has been connected. But those are definitely the events that occur *after* the device has been connected.

A simple workaround is to just sleep for 100 milliseconds before listing the files in `/dev/input`, but perhaps there's a more elegant fix.

If you can't reproduce this issue on your machine, I'm happy to help, let me know if you want me test any patches.

**Comment by k0kubun** [comment]
Hi, thank you for your report. Could you try this branch https://github.com/k0kubun/xremap/pull/61 and see if it helps your situation?

**Comment by nroi** [comment]
Thanks for looking into it. Unfortunately, the issue is still there on branch `ungrab-devices`. I haven't noticed any difference in the logs.

**Comment by k0kubun** [comment]
I see.

Just to let me have the full contexts, let me ask another question first. Are the logs you pasted in your first comment include everything from the moment when xremap started? If not, could you start the logs from the beginning so that I can see as many logs as possible?

**Comment by k0kubun** [comment]
JFYI, I use Ubuntu 20.04 w/ Xorg GNOME and can't reproduce this. What environment do you use?

> I really have no idea why I get so many ATTRIB events, and why it closes the file multiple times after the device has been connected. But those are definitely the events that occur after the device has been connected.
>
> A simple workaround is to just sleep for 100 milliseconds before listing the files in /dev/input

Can you see if the permission is denied right after the last `/dev/input/ MODIFY event3`? Do you need 100 milliseconds even after the last MODIFY event?

**Comment by k0kubun** [comment]
Also, could you try this branch? https://github.com/k0kubun/xremap/pull/62

**Comment by nroi** [comment]
I use ArchLinux with Kernel 5.15.11, Xorg and i3 as window manager.

> JFYI, I use Ubuntu 20.04 w/ Xorg GNOME and can't reproduce this.

That's strange, I hope it's not due to something that I messed up myself on my own system. I'm going to try it with a different window manager or distribution just to check which part of the environment is responsible for this. I will let you know tomorrow or the day after that.

> Can you see if the permission is denied right after the last /dev/input/ MODIFY event3? Do you need 100 milliseconds even after the last MODIFY event?

I'll try to find that out, too!

I've tried the branch #62, but it fails with this output:

```
Found a removed device. Reselecting devices.
Selecting devices from the following list:
------------------------------------------------------------------------------
/dev/input/event0 : Power Button
/dev/input/event1 : Power Button
/dev/input/event5 : PixArt USB Optical Mouse
/dev/input/event6 : UVC Camera (046d:0825)
/dev/input/event7 : HDA ATI HDMI HDMI/DP,pcm=3
/dev/input/event8 : HDA ATI HDMI HDMI/DP,pcm=7
/dev/input/event9 : HDA ATI HDMI HDMI/DP,pcm=8
/dev/input/event10: HDA ATI HDMI HDMI/DP,pcm=9
/dev/input/event11: HDA ATI HDMI HDMI/DP,pcm=10
/dev/input/event12: HDA ATI HDMI HDMI/DP,pcm=11
/dev/input/event13: HD-Audio Generic Front Mic
/dev/input/event14: HD-Audio Generic Rear Mic
/dev/input/event15: HD-Audio Generic Line
/dev/input/event16: HD-Audio Generic Line Out Front
/dev/input/event17: HD-Audio Generic Line Out Surround
/dev/input/event18: HD-Audio Generic Line Out CLFE
/dev/input/event19: HD-Audio Generic Front Headphone
/dev/input/event20: USB Keyboard System Control
/dev/input/event21: BurrBrown from Texas Instruments USB AUDIO  CODEC
------------------------------------------------------------------------------
Selected keyboards automatically since --device options weren't specified:
------------------------------------------------------------------------------
Failed to prepare input devices: No device was selected!
```

>  Are the logs you pasted in your first comment include everything from the moment when xremap started? If not, could you start the logs from the beginning so that I can see as many logs as possible?

Here's the full log, but I don't think it includes anything interesting that I've skipped:

```
Path "/dev/input/event4" cannot be opened: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
Selecting devices from the following list:
------------------------------------------------------------------------------
/dev/input/event0 : Power Button
/dev/input/event1 : Power Button
/dev/input/event2 : USB Keyboard Consumer Control
/dev/input/event3 : USB Keyboard
/dev/input/event5 : PixArt USB Optical Mouse
/dev/input/event6 : UVC Camera (046d:0825)
/dev/input/event7 : HDA ATI HDMI HDMI/DP,pcm=3
/dev/input/event8 : HDA ATI HDMI HDMI/DP,pcm=7
/dev/input/event9 : HDA ATI HDMI HDMI/DP,pcm=8
/dev/input/event10: HDA ATI HDMI HDMI/DP,pcm=9
/dev/input/event11: HDA ATI HDMI HDMI/DP,pcm=10
/dev/input/event12: HDA ATI HDMI HDMI/DP,pcm=11
/dev/input/event13: HD-Audio Generic Front Mic
/dev/input/event14: HD-Audio Generic Rear Mic
/dev/input/event15: HD-Audio Generic Line
/dev/input/event16: HD-Audio Generic Line Out Front
/dev/input/event17: HD-Audio Generic Line Out Surround
/dev/input/event18: HD-Audio Generic Line Out CLFE
/dev/input/event19: HD-Audio Generic Front Headphone
/dev/input/event20: USB Keyboard System Control
/dev/input/event21: BurrBrown from Texas Instruments USB AUDIO  CODEC
------------------------------------------------------------------------------
Selected keyboards automatically since --device options weren't specified:
------------------------------------------------------------------------------
/dev/input/event3 : USB Keyboard
------------------------------------------------------------------------------
Received events from inotify: Err(EAGAIN)

[ KEYBOARD DISCONNECTED AND RECONNECTED ]

Received events from inotify: Ok([InotifyEvent { wd: WatchDescriptor { wd: 1 }, mask: IN_ATTRIB, cookie: 0, name: Some("event3") }])
Found a removed device. Reselecting devices.
Path "/dev/input/event2" cannot be opened: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
Selecting devices from the following list:
------------------------------------------------------------------------------
/dev/input/event0 : Power Button
/dev/input/event1 : Power Button
/dev/input/event5 : PixArt USB Optical Mouse
/dev/input/event6 : UVC Camera (046d:0825)
/dev/input/event7 : HDA ATI HDMI HDMI/DP,pcm=3
/dev/input/event8 : HDA ATI HDMI HDMI/DP,pcm=7
/dev/input/event9 : HDA ATI HDMI HDMI/DP,pcm=8
/dev/input/event10: HDA ATI HDMI HDMI/DP,pcm=9
/dev/input/event11: HDA ATI HDMI HDMI/DP,pcm=10
/dev/input/event12: HDA ATI HDMI HDMI/DP,pcm=11
/dev/input/event13: HD-Audio Generic Front Mic
/dev/input/event14: HD-Audio Generic Rear Mic
/dev/input/event15: HD-Audio Generic Line
/dev/input/event16: HD-Audio Generic Line Out Front
/dev/input/event17: HD-Audio Generic Line Out Surround
/dev/input/event18: HD-Audio Generic Line Out CLFE
/dev/input/event19: HD-Audio Generic Front Headphone
/dev/input/event21: BurrBrown from Texas Instruments USB AUDIO  CODEC
------------------------------------------------------------------------------
Selected keyboards automatically since --device options weren't specified:
------------------------------------------------------------------------------
warning: No device was selected, but --watch is waiting for new devices.
------------------------------------------------------------------------------
Received events from inotify: Ok([InotifyEvent { wd: WatchDescriptor { wd: 1 }, mask: IN_CREATE, cookie: 0, name: Some("event3") }, InotifyEvent { wd: WatchDescriptor { wd: 1 }, mask: IN_ATTRIB, cookie: 0, name: Some("event3") }])
Detected device changes. Reselecting devices.
Path "/dev/input/event4" cannot be opened: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
Path "/dev/input/event2" cannot be opened: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
Path "/dev/input/event3" cannot be opened: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
Selecting devices from the following list:
------------------------------------------------------------------------------
/dev/input/event0 : Power Button
/dev/input/event1 : Power Button
/dev/input/event5 : PixArt USB Optical Mouse
/dev/input/event6 : UVC Camera (046d:0825)
/dev/input/event7 : HDA ATI HDMI HDMI/DP,pcm=3
/dev/input/event8 : HDA ATI HDMI HDMI/DP,pcm=7
/dev/input/event9 : HDA ATI HDMI HDMI/DP,pcm=8
/dev/input/event10: HDA ATI HDMI HDMI/DP,pcm=9
/dev/input/event11: HDA ATI HDMI HDMI/DP,pcm=10
/dev/input/event12: HDA ATI HDMI HDMI/DP,pcm=11
/dev/input/event13: HD-Audio Generic Front Mic
/dev/input/event14: HD-Audio Generic Rear Mic
/dev/input/event15: HD-Audio Generic Line
/dev/input/event16: HD-Audio Generic Line Out Front
/dev/input/event17: HD-Audio Generic Line Out Surround
/dev/input/event18: HD-Audio Generic Line Out CLFE
/dev/input/event19: HD-Audio Generic Front Headphone
/dev/input/event21: BurrBrown from Texas Instruments USB AUDIO  CODEC
------------------------------------------------------------------------------
Selected keyboards automatically since --device options weren't specified:
------------------------------------------------------------------------------
warning: No device was selected, but --watch is waiting for new devices.
------------------------------------------------------------------------------
```

**Comment by k0kubun** [comment]
Thank you for the information. To dissect the situation further, could you also try starting xremap with sudo? Does the "Permission denied" also happen for the root user? And out of curiosity, do you know what is event4 that is always "Permission denied"? Is that just xremap itself?

**Comment by nroi** [comment]
>  could you also try starting xremap with sudo? Does the "Permission denied" also happen for the root user?

It works as intended when I run it as sudo, without the permission denied error.

> do you know what is event4 that is always "Permission denied"? Is that just xremap itself?

I don't know, and I'm not sure how to find that out. event4 is already present when I boot the system, before I have started xremap, so it's probably something else.

**Comment by k0kubun** [comment]
> It works as intended when I run it as sudo, without the permission denied error.

Nice. Out of curiosity, would that work as a last-resort workaround for you? I'm actually planning to update README.md again to also mention the sudo method since I get problem reports from people using environments like Arch.

Also, could you take a look at what permission and owner the problematic device has when you get "Permission denied"? Is that changed when you wait for 100ms?

> I'm not sure how to find that out.

I think you can use `cat /proc/bus/input/devices` to see what event4 is.

**Comment by nroi** [comment]
`event4` is `USB Keyboard System Control` in my case. So my keyboard seems to have multiple event entries in `/dev/input`.

I did a few more tests:
* External keyboard on my laptop running an Ubuntu-based distribution: Works as intended.
* A clean install of Ubuntu 20.04 with Xorg on my desktop system (the system that also has Arch Linux installed, where the issue occurs): Does not work. Same issue as with Arch Linux.
* The same Ubuntu system with a different keyboard: Same issue.

So, I guess that means that the issue is not ArchLinux, and also not the keyboard.

> Nice. Out of curiosity, would that work as a last-resort workaround for you?

Yes, that works for me. Feel free to close this issue.

**Comment by k0kubun** [comment]
> So, I guess that means that the issue is not ArchLinux, and also not the keyboard.

From your observation, it seems to me like the problem comes from either your laptop <-> your desktop or the Ubuntu-based distribution <-> Ubuntu 20.04 with Xorg. Since it works on my Ubuntu 20.04 environment, it's probably not Ubuntu or Linux kernel versions but something installed on top of it.

> Yes, that works for me. Feel free to close this issue.

Thank you. I'll close this once I resurrect a document for sudo usages, and I guess I'll need to suggest this for such users until I get an environment that reproduces the issue. I'll probably try installing Arch Linux when I work on it.

Also, if you're interested in helping me find the reason, these questions are still valid:

> could you take a look at what permission and owner the problematic device has when you get "Permission denied"? Is that changed when you wait for 100ms?

**Comment by k0kubun** [comment]
In particular, I expect that `/etc/input/eventX` is owned by `input` group, read and write are permitted to the group, and your normal user is in the input group.

**Comment by nroi** [comment]
> could you take a look at what permission and owner the problematic device has when you get "Permission denied"? Is that changed when you wait for 100ms?

This is the output of `ls -ls /dev/input/event3` when the permission denied error occurs:

```
0 crw------- 1 root root 13, 67 Jan 13 21:34 /dev/input/event3
```

This is the output a few seconds after the permission denied error:

```
0 crw-rw---- 1 root input 13, 67 Jan 13 21:34 /dev/input/event3
```

**Comment by k0kubun** [comment]
Thank you. Could you try this branch then https://github.com/k0kubun/xremap/pull/64?

**Comment by nroi** [comment]
I tried #64, but even without reconnecting any device, it seems to run in an infinite loop. Here's an excerpt of the log.

```
Selected keyboards automatically since --device options weren't specified:
------------------------------------------------------------------------------
/dev/input/event2 : USB Keyboard
------------------------------------------------------------------------------
Detected device changes. Reselecting devices.
^[Selecting devices from the following list:
------------------------------------------------------------------------------
/dev/input/event0 : Power Button
/dev/input/event1 : Power Button
/dev/input/event2 : USB Keyboard
/dev/input/event3 : PixArt USB Optical Mouse
/dev/input/event4 : BurrBrown from Texas Instruments USB AUDIO  CODEC
/dev/input/event5 : HDA ATI HDMI HDMI/DP,pcm=3
/dev/input/event6 : HDA ATI HDMI HDMI/DP,pcm=7
/dev/input/event7 : HDA ATI HDMI HDMI/DP,pcm=8
/dev/input/event8 : HDA ATI HDMI HDMI/DP,pcm=9
/dev/input/event9 : HDA ATI HDMI HDMI/DP,pcm=10
/dev/input/event10: HDA ATI HDMI HDMI/DP,pcm=11
/dev/input/event11: HD-Audio Generic Front Mic
/dev/input/event12: HD-Audio Generic Rear Mic
/dev/input/event13: HD-Audio Generic Line
/dev/input/event14: HD-Audio Generic Line Out Front
/dev/input/event15: HD-Audio Generic Line Out Surround
/dev/input/event16: HD-Audio Generic Line Out CLFE
/dev/input/event17: HD-Audio Generic Front Headphone
/dev/input/event18: UVC Camera (046d:0825)
/dev/input/event19: USB Keyboard Consumer Control
/dev/input/event20: USB Keyboard System Control
------------------------------------------------------------------------------
Selected keyboards automatically since --device options weren't specified:
------------------------------------------------------------------------------
/dev/input/event2 : USB Keyboard
------------------------------------------------------------------------------
Detected device changes. Reselecting devices.
^[Selecting devices from the following list:
------------------------------------------------------------------------------
/dev/input/event0 : Power Button
/dev/input/event1 : Power Button
/dev/input/event2 : USB Keyboard
/dev/input/event3 : PixArt USB Optical Mouse
/dev/input/event4 : BurrBrown from Texas Instruments USB AUDIO  CODEC
/dev/input/event5 : HDA ATI HDMI HDMI/DP,pcm=3
/dev/input/event6 : HDA ATI HDMI HDMI/DP,pcm=7
/dev/input/event7 : HDA ATI HDMI HDMI/DP,pcm=8
/dev/input/event8 : HDA ATI HDMI HDMI/DP,pcm=9
/dev/input/event9 : HDA ATI HDMI HDMI/DP,pcm=10
/dev/input/event10: HDA ATI HDMI HDMI/DP,pcm=11
/dev/input/event11: HD-Audio Generic Front Mic
/dev/input/event12: HD-Audio Generic Rear Mic
/dev/input/event13: HD-Audio Generic Line
/dev/input/event14: HD-Audio Generic Line Out Front
/dev/input/event15: HD-Audio Generic Line Out Surround
/dev/input/event16: HD-Audio Generic Line Out CLFE
/dev/input/event17: HD-Audio Generic Front Headphone
/dev/input/event18: UVC Camera (046d:0825)
/dev/input/event19: USB Keyboard Consumer Control
/dev/input/event20: USB Keyboard System Control
------------------------------------------------------------------------------
Selected keyboards automatically since --device options weren't specified:
------------------------------------------------------------------------------
/dev/input/event2 : USB Keyboard
------------------------------------------------------------------------------
Detected device changes. Reselecting devices.
Selecting devices from the following list:
------------------------------------------------------------------------------
/dev/input/event0 : Power Button
/dev/input/event1 : Power Button
/dev/input/event2 : USB Keyboard
/dev/input/event3 : PixArt USB Optical Mouse
/dev/input/event4 : BurrBrown from Texas Instruments USB AUDIO  CODEC
/dev/input/event5 : HDA ATI HDMI HDMI/DP,pcm=3
/dev/input/event6 : HDA ATI HDMI HDMI/DP,pcm=7
/dev/input/event7 : HDA ATI HDMI HDMI/DP,pcm=8
/dev/input/event8 : HDA ATI HDMI HDMI/DP,pcm=9
/dev/input/event9 : HDA ATI HDMI HDMI/DP,pcm=10
/dev/input/event10: HDA ATI HDMI HDMI/DP,pcm=11
/dev/input/event11: HD-Audio Generic Front Mic
/dev/input/event12: HD-Audio Generic Rear Mic
/dev/input/event13: HD-Audio Generic Line
/dev/input/event14: HD-Audio Generic Line Out Front
/dev/input/event15: HD-Audio Generic Line Out Surround
/dev/input/event16: HD-Audio Generic Line Out CLFE
/dev/input/event17: HD-Audio Generic Front Headphone
/dev/input/event18: UVC Camera (046d:0825)
/dev/input/event19: USB Keyboard Consumer Control
/dev/input/event20: USB Keyboard System Control
------------------------------------------------------------------------------
Selected keyboards automatically since --device options weren't specified:
------------------------------------------------------------------------------
/dev/input/event2 : USB Keyboard
------------------------------------------------------------------------------
Detected device changes. Reselecting devices.
Selecting devices from the following list:
------------------------------------------------------------------------------
/dev/input/event0 : Power Button
/dev/input/event1 : Power Button
/dev/input/event2 : USB Keyboard
/dev/input/event3 : PixArt USB Optical Mouse
/dev/input/event4 : BurrBrown from Texas Instruments USB AUDIO  CODEC
/dev/input/event5 : HDA ATI HDMI HDMI/DP,pcm=3
/dev/input/event6 : HDA ATI HDMI HDMI/DP,pcm=7
/dev/input/event7 : HDA ATI HDMI HDMI/DP,pcm=8
/dev/input/event8 : HDA ATI HDMI HDMI/DP,pcm=9
/dev/input/event9 : HDA ATI HDMI HDMI/DP,pcm=10
/dev/input/event10: HDA ATI HDMI HDMI/DP,pcm=11
/dev/input/event11: HD-Audio Generic Front Mic
/dev/input/event12: HD-Audio Generic Rear Mic
/dev/input/event13: HD-Audio Generic Line
/dev/input/event14: HD-Audio Generic Line Out Front
/dev/input/event15: HD-Audio Generic Line Out Surround
/dev/input/event16: HD-Audio Generic Line Out CLFE
/dev/input/event17: HD-Audio Generic Front Headphone
/dev/input/event18: UVC Camera (046d:0825)
/dev/input/event19: USB Keyboard Consumer Control
/dev/input/event20: USB Keyboard System Control
------------------------------------------------------------------------------
Selected keyboards automatically since --device options weren't specified:
------------------------------------------------------------------------------
/dev/input/event2 : USB Keyboard
------------------------------------------------------------------------------
Detected device changes. Reselecting devices.
Selecting devices from the following list:
------------------------------------------------------------------------------
/dev/input/event0 : Power Button
/dev/input/event1 : Power Button
/dev/input/event2 : USB Keyboard
/dev/input/event3 : PixArt USB Optical Mouse
/dev/input/event4 : BurrBrown from Texas Instruments USB AUDIO  CODEC
/dev/input/event5 : HDA ATI HDMI HDMI/DP,pcm=3
/dev/input/event6 : HDA ATI HDMI HDMI/DP,pcm=7
/dev/input/event7 : HDA ATI HDMI HDMI/DP,pcm=8
/dev/input/event8 : HDA ATI HDMI HDMI/DP,pcm=9
/dev/input/event9 : HDA ATI HDMI HDMI/DP,pcm=10
/dev/input/event10: HDA ATI HDMI HDMI/DP,pcm=11
/dev/input/event11: HD-Audio Generic Front Mic
/dev/input/event12: HD-Audio Generic Rear Mic
/dev/input/event13: HD-Audio Generic Line
/dev/input/event14: HD-Audio Generic Line Out Front
/dev/input/event15: HD-Audio Generic Line Out Surround
/dev/input/event16: HD-Audio Generic Line Out CLFE
/dev/input/event17: HD-Audio Generic Front Headphone
/dev/input/event18: UVC Camera (046d:0825)
/dev/input/event19: USB Keyboard Consumer Control
/dev/input/event20: USB Keyboard System Control
------------------------------------------------------------------------------
Selected keyboards automatically since --device options weren't specified:
```

**Comment by k0kubun** [comment]
Oops, sorry about that. My suspicion was that we're losing the event of changing `crw-------` to `crw-rw----` due to a race condition, and I applied a change to make sure we keep all events, but apparently it captured too many events. I'll comment here again when my next attempt is ready.

**Comment by k0kubun** [comment]
Another person volunteered to make an improvement here.

@nroi Could you see if this branch https://github.com/k0kubun/xremap/pull/78 reloads devices correctly without sudo?

**Comment by nroi** [comment]
@k0kubun looks good, the mapping is correctly applied when I disconnect and reconnect my keyboard, no sudo required. :+1: 

When I reconnect my keyboard, the following error message appears:

```
Failed to grab device 'USB Keyboard' at '/dev/input/event2' due to: Device or resource busy (os error 16)
```

But like I said, everything seems to work as intended.

**Comment by k0kubun** [comment]
Thank you for confirming it. I'm glad to hear that this problem is now gone. The new version has been released as v0.3.0.
